### PR TITLE
Add Listening coach with edit and add

### PR DESCRIPTION
**Pull request by shuliakleonid**
CPT2108152108-1197x952

### CODE REVIEW

File: src/app/constants/data-constants.ts [hunk 1]
```diff
@@ -44,36 +44,263 @@ export enum GrammarAnswers {
...
-  answers: [
+  answers: emptyAnswers(),
+};
+
+export const emptyTopic: CoachEmptyTopic = {
+  level: 1,
+  topicName: '',
+};
+
+// export const emptyQuestionsListening = () => {
```
Comment (RomPavel) [inline src/app/constants/data-constants.ts:65]: why not functions

File: src/app/models/coach-profile-editor/coach-profile-editor.component.ts [hunk 2]
```diff
@@ -42,6 +49,8 @@ export class CoachProfileEditorComponent implements OnInit {
       this.selectedTab = CoachEditorTabs.grammar;
```
Comment (RomPavel) [inline src/app/models/coach-profile-editor/coach-profile-editor.component.ts:49]: there is a bug, when you switch from 2nd or 3rd tab to first one

File: src/app/models/common-test/common-test.component.ts [hunk 3]
```diff
@@ -30,7 +30,7 @@ export class CommonTestComponent implements OnInit {
     this.testStoreService.createTestContent();
 
     this.grammar$ = this.test$.pipe(map((test) => test?.grammarQuestions || null));
-    this.listening$ = this.test$.pipe(map((test) => test?.audition.questions || null));
+    // this.listening$ = this.test$.pipe(map((test) => test?.audition.questions || null));
```
Comment (RomPavel) [inline src/app/models/common-test/common-test.component.ts:33]: why

File: src/app/services/coach-listening-http.service.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+  getListeningList() {
+    return this.http.get<ResponseGetAll<CoachListening>>(ListeningApiUrl);
+  }
+
+  getListening(id: string) {
+    return this.http.get<EditionCoachListening>(`${ListeningApiUrl}/${id}`);
+  }
+
+  updateListening(topic: UpdateCoachListening) {
```
Comment (RomPavel) [inline src/app/services/coach-listening-http.service.ts:29]: topic -> listening

File: src/app/services/store/coach-listening-store.service.ts [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+import { map } from 'rxjs/operators';
+import { TableData } from '../../interfaces/question-answer';
+import { CoachListeningHttpService } from '../coach-listening-http.service';
+import { EditionCoachListening, UpdateCoachListening } from '../../interfaces/audition';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class CoachListeningStoreService {
+  readonly listen$ = new Subject<EditionCoachListening>();
```
Comment (RomPavel) [inline src/app/services/store/coach-listening-store.service.ts:12]: listen -> listening

File: src/app/services/store/coach-listening-store.service.ts [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+  readonly listens$ = new Subject<TableData[]>();
```
Comment (RomPavel) [inline src/app/services/store/coach-listening-store.service.ts:14]: same here

File: src/app/models/coach-profile-editor/listening-adding-editing-dialog/listening-adding-editing-dialog.component.html [hunk 7]
```diff
@@ -0,0 +1,55 @@
+<div class="main">
+  <h1>{{ data.isEdit ? 'Edit Listening' : 'Add Listening' }}</h1>
+  <div>
+    <input type="file" (change)="handleFileInput($event)" />
```
Comment (RomPavel) [inline src/app/models/coach-profile-editor/listening-adding-editing-dialog/listening-adding-editing-dialog.component.html:4]: if download file endpoint works, than we need to show previous audio (better with audio tag)

File: src/app/models/common-test/common-test.component.ts (hunk 3 above)
Comment (shuliakleonid) [reply src/app/models/common-test/common-test.component.ts]: done, my fault

File: src/app/services/coach-listening-http.service.ts (hunk 4 above)
Comment (shuliakleonid) [reply src/app/services/coach-listening-http.service.ts:29]: done

File: src/app/models/coach-profile-editor/listening-adding-editing-dialog/listening-adding-editing-dialog.component.html [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+  </div>
```
Comment (RomPavel) [inline src/app/models/coach-profile-editor/listening-adding-editing-dialog/listening-adding-editing-dialog.component.html:5]: we need file input to be styled, at least like button white button

File: src/app/constants/data-constants.ts (hunk 1 above)
Comment (shuliakleonid) [reply src/app/constants/data-constants.ts:65]: done

File: src/app/models/coach-profile-editor/coach-profile-editor.component.ts (hunk 2 above)
Comment (shuliakleonid) [reply src/app/models/coach-profile-editor/coach-profile-editor.component.ts:49]: done

File: src/app/services/store/coach-listening-store.service.ts (hunk 5 above)
Comment (shuliakleonid) [reply src/app/services/store/coach-listening-store.service.ts:12]: done

File: src/app/services/store/coach-listening-store.service.ts (hunk 6 above)
Comment (shuliakleonid) [reply src/app/services/store/coach-listening-store.service.ts:14]: done

**Comment by RomPavel** [comment]
Modal should not automatically scroll to end

File: src/app/models/coach-profile-editor/listening-adding-editing-dialog/listening-adding-editing-dialog.component.html [hunk 9]
```diff
@@ -0,0 +1,59 @@
+<div class="main">
+  <h1>{{ data.isEdit ? 'Edit Listening' : 'Add Listening' }}</h1>
+  <div>
+    <label class="load-file">
+      <input type="file" (change)="handleFileInput($event)" />
```
Comment (RomPavel) [inline src/app/models/coach-profile-editor/listening-adding-editing-dialog/listening-adding-editing-dialog.component.html:5]: we should specify files accept types
image

File: src/app/models/coach-profile-editor/listening-adding-editing-dialog/listening-adding-editing-dialog.component.html [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+      <span>Custom Upload</span>
```
Comment (RomPavel) [inline src/app/models/coach-profile-editor/listening-adding-editing-dialog/listening-adding-editing-dialog.component.html:6]: What does Custom means

File: src/app/constants/data-constants.ts (hunk 1 above)
Comment (shuliakleonid) [reply src/app/constants/data-constants.ts]: done

File: src/app/services/store/coach-listening-store.service.ts (hunk 6 above)
Comment (shuliakleonid) [reply src/app/services/store/coach-listening-store.service.ts]: done

File: src/app/models/coach-profile-editor/listening-adding-editing-dialog/listening-adding-editing-dialog.component.html (hunk 7 above)
Comment (shuliakleonid) [reply src/app/models/coach-profile-editor/listening-adding-editing-dialog/listening-adding-editing-dialog.component.html]: done

File: src/app/models/coach-profile-editor/editor-table/editor-table.component.ts [hunk 11]
```diff
@@ -116,9 +118,14 @@ export class EditorTableComponent implements AfterViewInit, OnChanges, OnInit {
...
     } else if (this.selectTab === CoachEditorTabs.audition) {
-      this.dialog.open(EditListeningDialogComponent, {
-        autoFocus: false,
-        disableClose: true,
+      this.coachListening.getListening(row.id);
+      this.coachListening.listening$.pipe(take(1)).subscribe((listen) => {
+        if (listen !== null) {
+          this.dialog.open(ListeningAddingEditingDialogComponent, {
+            data: { ...listen, isEdit: true },
+            disableClose: true,
```
Comment (RomPavel) [inline src/app/models/coach-profile-editor/editor-table/editor-table.component.ts:126]: Modal should not automatically scroll to end

File: src/app/models/coach-profile-editor/listening-adding-editing-dialog/listening-adding-editing-dialog.component.html (hunk 7 above)
Comment (RomPavel) [reply src/app/models/coach-profile-editor/listening-adding-editing-dialog/listening-adding-editing-dialog.component.html]: I see that you show audio on upload, but you dont show audio when edit listening

File: src/app/models/coach-profile-editor/listening-adding-editing-dialog/listening-adding-editing-dialog.component.html (hunk 8 above)
Comment (shuliakleonid) [reply src/app/models/coach-profile-editor/listening-adding-editing-dialog/listening-adding-editing-dialog.component.html:9]: done

File: src/app/models/coach-profile-editor/listening-adding-editing-dialog/listening-adding-editing-dialog.component.html (hunk 10 above)
Comment (shuliakleonid) [reply src/app/models/coach-profile-editor/listening-adding-editing-dialog/listening-adding-editing-dialog.component.html]: done

File: src/app/models/coach-profile-editor/listening-adding-editing-dialog/listening-adding-editing-dialog.component.html (hunk 9 above)
Comment (shuliakleonid) [reply src/app/models/coach-profile-editor/listening-adding-editing-dialog/listening-adding-editing-dialog.component.html]: done

File: src/app/services/store/coach-listening-store.service.ts (hunk 5 above)
Comment (shuliakleonid) [reply src/app/services/store/coach-listening-store.service.ts]: done

File: src/app/models/coach-profile-editor/coach-profile-editor.component.ts (hunk 2 above)
Comment (shuliakleonid) [reply src/app/models/coach-profile-editor/coach-profile-editor.component.ts:48]: done

File: src/app/models/coach-profile-editor/listening-adding-editing-dialog/listening-adding-editing-dialog.component.html (hunk 7 above)
Comment (shuliakleonid) [reply src/app/models/coach-profile-editor/listening-adding-editing-dialog/listening-adding-editing-dialog.component.html]: backend don`t receive path or file to us

File: src/app/models/coach-profile-editor/editor-table/editor-table.component.ts (hunk 11 above)
Comment (shuliakleonid) [reply src/app/models/coach-profile-editor/editor-table/editor-table.component.ts:126]: done

Comment (shuliakleonid) [reply src/app/models/coach-profile-editor/editor-table/editor-table.component.ts:126]: done

**Review by casterweb — APPROVED** [review]

**Pull request merged**